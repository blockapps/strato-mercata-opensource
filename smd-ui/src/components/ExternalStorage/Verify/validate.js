export default function validate(values) {
  const errors = {};

  if (!values.contractAddress) {
    errors.contractAddress = "Can't be blank";
  }

  return errors;
}