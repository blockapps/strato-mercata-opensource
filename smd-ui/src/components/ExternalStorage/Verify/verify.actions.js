export const OPEN_VERIFY_MODAL = 'OPEN_VERIFY_MODAL';
export const CLOSE_VERIFY_MODAL = 'CLOSE_VERIFY_MODAL';
export const VERIFY_DOCUMENT_REQUEST = 'VERIFY_DOCUMENT_REQUEST';
export const VERIFY_DOCUMENT_SUCCESS = 'VERIFY_DOCUMENT_SUCCESS';
export const VERIFY_DOCUMENT_FAILURE = 'VERIFY_DOCUMENT_FAILURE';
export const RESET_ERROR = 'RESET_ERROR';

export const openVerifyModal = function () {
  return {
    type: OPEN_VERIFY_MODAL
  }
}

export const closeVerifyModal = function () {
  return {
    type: CLOSE_VERIFY_MODAL
  }
}

export const verifyDocumentRequest = function (contractAddress) {
  return {
    type: VERIFY_DOCUMENT_REQUEST,
    contractAddress
  }
}

export const verifyDocumentSuccess = function (data) {
  return {
    type: VERIFY_DOCUMENT_SUCCESS,
    data
  }
}

export const verifyDocumentFailure = function (error) {
  return {
    type: VERIFY_DOCUMENT_FAILURE,
    error
  }
}

export const resetError = function () {
  return {
    type: RESET_ERROR
  }
}