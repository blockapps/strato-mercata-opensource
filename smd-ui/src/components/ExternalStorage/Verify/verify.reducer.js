import {
  OPEN_VERIFY_MODAL, CLOSE_VERIFY_MODAL, VERIFY_DOCUMENT_REQUEST, VERIFY_DOCUMENT_SUCCESS, VERIFY_DOCUMENT_FAILURE, RESET_ERROR
} from "./verify.actions";

const initialState = {
  isOpen: false,
  verifyDocument: null,
  error: null
}

const reducer = function (state = initialState, action) {
  switch (action.type) {
    case OPEN_VERIFY_MODAL:
      return {
        ...state,
        isOpen: true
      }
    case CLOSE_VERIFY_MODAL:
      return {
        ...state,
        isOpen: false,
        verifyDocument: null
      }
    case VERIFY_DOCUMENT_REQUEST:
      return {
        ...state,
        isLoading: true
      }
    case VERIFY_DOCUMENT_SUCCESS:
      return {
        ...state,
        isLoading: false,
        verifyDocument: action.data,
        error: null
      }
    case VERIFY_DOCUMENT_FAILURE:
      return {
        ...state,
        isLoading: false,
        verifyDocument: null,
        error: action.error
      }
    case RESET_ERROR:
      return {
        ...state,
        error: null
      }
    default:
      return state;
  }
}

export default reducer;