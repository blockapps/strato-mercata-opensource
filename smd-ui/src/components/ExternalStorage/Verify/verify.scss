.verify-result ul {
  list-style-type: none;
  padding: 0px !important;
}

.verify-result li {
  color: lightgreen;
}

.verify-result li:before {
  font-family: 'FontAwesome';
  content: '\F00C';
  margin: 0 5px 0 0px;
  color: lightgreen;
  font-size: 20px;
}

.verify-dialog .content-margin {
  margin: 10px 10px;
}

.verify-dialog {
  width: 580px;
}