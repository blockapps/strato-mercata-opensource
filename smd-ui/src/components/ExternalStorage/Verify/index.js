import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Dialog, Button, Intent } from '@blueprintjs/core';
import { Field, reduxForm } from 'redux-form';
import mixpanelWrapper from '../../../lib/mixpanelWrapper';
import validate from './validate';
import { closeVerifyModal, verifyDocumentRequest, resetError } from './verify.actions';
import { parseDateFromTimestamp } from '../../../lib/dateUtils';
import { toasts } from '../../Toasts';

import './verify.css';

class Verify extends Component {

  constructor(props) {
    super(props);
    this.state = { errors: null }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.verifyError) {
      toasts.show({ message: nextProps.verifyError });
      this.props.resetError();
    }
  }

  submit = (values) => {
    let errors = validate(values);
    this.setState({ errors });

    if (!Object.values(errors).length) {
      this.props.verifyDocumentRequest(values.contractAddress);
    }
  }

  errorMessageFor(fieldName) {
    if (this.state.errors && this.state.errors[fieldName]) {
      return this.state.errors[fieldName];
    }
    return null;
  }

  renderForm() {
    return (
      <div className="pt-dialog-body upload-form">

        <div className="row">
          <div className="col-sm-4 text-right">
            <label className="pt-label smd-pad-4">
              Contract Address
            </label>
          </div>
          <div className="col-sm-8 smd-pad-4">
            <Field
              name="contractAddress"
              component="input"
              type="text"
              placeholder="Contract Address"
              className="pt-input form-width"
              tabIndex="1"
              required
            />
            <br /><span className="error-text">{this.errorMessageFor('contractAddress')}</span>
          </div>
        </div>

        <div className="pt-dialog-footer">
          <div className="pt-dialog-footer-actions button-center smd-margin-8">
            <Button
              intent={Intent.PRIMARY}
              onClick={this.props.handleSubmit(this.submit)}
              disabled={this.props.isLoading}
              text="Attest"
            />
          </div>
        </div>
      </div>
    );
  }

  renderSuccess(data) {
    const signers = data.signers.map((value, key) => {
      return (
        <li key={key}> {value} </li>
      )
    })

    return (
      <div>
        <div className="pt-dialog-body">

          <div className="row content-margin">
            <div className="col-sm-3">
              <label> URI </label>
            </div>
            <div className="col-sm-9">
              <label> {data.uri} </label>
            </div>
          </div>

          <div className="row content-margin">
            <div className="col-sm-3">
              <label> Date Uploaded </label>
            </div>
            <div className="col-sm-9">
              <label> { parseDateFromTimestamp(data.timeStamp) } </label>
            </div>
          </div>

          <div className="row content-margin verify-result">
            <div className="col-sm-3">
              <label> Signatures </label>
            </div>
            <div className="col-sm-9">
              <ul> {signers} </ul>
            </div>
          </div>

        </div>

        <div className="pt-dialog-footer">
          <div className="pt-dialog-footer-actions button-center">
            <Button
              intent={Intent.PRIMARY}
              onClick={() => this.closeModal()}
              text="Close"
            />
          </div>
        </div>
      </div>
    )
  }

  closeModal() {
    this.props.closeVerifyModal();
    this.props.reset();
  }

  render() {
    const result = this.props.verifyDocument;
    return (
      <div>
        <form>
          <Dialog
            isOpen={this.props.isOpen}
            onClose={() => {
              mixpanelWrapper.track('close_verify_modal');
              this.closeModal();
            }}
            iconName={result ? 'saved' : 'pt-icon-info-sign'}
            title={result ? 'Valid Resource' : 'Verify'}
            className="pt-dark verify-dialog"
          >
            {result ? this.renderSuccess(result) : this.renderForm()}
          </Dialog>
        </form>
      </div>
    )
  }
}

export function mapStateToProps(state) {
  return {
    isOpen: state.verify.isOpen,
    isLoading: state.verify.isLoading,
    verifyDocument: state.verify.verifyDocument,
    verifyError: state.verify.error
  };
}

const formed = reduxForm({ form: 'verify-form' })(Verify);
const connected = connect(
  mapStateToProps,
  {
    verifyDocumentRequest,
    closeVerifyModal,
    resetError
  }
)(formed);

export default withRouter(connected);
