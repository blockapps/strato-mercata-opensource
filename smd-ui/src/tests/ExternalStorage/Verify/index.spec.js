import React from 'react';
import Verify, { mapStateToProps } from '../../../components/ExternalStorage/Verify';
import { createStore, combineReducers } from 'redux';
import { reducer as formReducer } from 'redux-form';

describe('Verify: index', () => {
  let store

  beforeEach(() => {
    store = createStore(combineReducers({ form: formReducer }));
  });

  describe('render component', () => {

    test('without values', () => {
      const props = {
        isOpen: false,
        isLoading: false,
        verifyDocument: null,
        verifyError: 'error',
        closeVerifyModal: jest.fn(),
        reset: jest.fn(),
        resetError: jest.fn(),
        verifyDocumentRequest: jest.fn(),
        handleSubmit: jest.fn(),
        store: store
      };

      const wrapper = shallow(
        <Verify.WrappedComponent {...props} />
      ).dive().dive().dive();

      expect(wrapper.debug()).toMatchSnapshot();
    });

    test('with values', () => {
      const props = {
        isOpen: true,
        isLoading: false,
        verifyDocument: {
          "uri": "https://strato-external-storage.s3.amazonaws.com/1530182373708-widescreen.jpeg",
          "timeStamp": 1531721964,
          "signers": [
            "6e873015e8ff27d7c6d3ab5d1403a9df9ab420ad",
            "a51f27e78aef85a06631f0725f380001e0ae9fb6"
          ]
        },
        verifyError: 'error',
        closeVerifyModal: jest.fn(),
        reset: jest.fn(),
        resetError: jest.fn(),
        verifyDocumentRequest: jest.fn(),
        handleSubmit: jest.fn(),
        store: store
      };

      const wrapper = shallow(
        <Verify.WrappedComponent {...props} />
      ).dive().dive().dive();

      expect(wrapper.debug()).toMatchSnapshot();
    });

  });

  describe('form', () => {

    test('test', () => {
      const props = {
        isOpen: false,
        isLoading: false,
        verifyDocument: null,
        verifyError: 'error',
        closeVerifyModal: jest.fn(),
        reset: jest.fn(),
        resetError: jest.fn(),
        verifyDocumentRequest: jest.fn(),
        handleSubmit: jest.fn(),
        store: store
      };

      const wrapper = shallow(
        <Verify.WrappedComponent {...props} />
      ).dive().dive().dive();

      wrapper.find('Field').simulate('change', { target: { value: '6e873015e8ff27d7c6d3ab5d1403a9df9ab420ad' } });
      wrapper.find('Button').simulate('click');
      expect(props.handleSubmit).toHaveBeenCalled();
    });

    test('submit with values', () => {
      const props = {
        isOpen: false,
        isLoading: false,
        verifyDocument: null,
        verifyError: 'error',
        closeVerifyModal: jest.fn(),
        reset: jest.fn(),
        resetError: jest.fn(),
        verifyDocumentRequest: jest.fn(),
        handleSubmit: jest.fn(),
        store: store
      };

      const wrapper = shallow(
        <Verify.WrappedComponent {...props} />
      ).dive().dive().dive();

      wrapper.instance().submit({ contractAddress: '6e873015e8ff27d7c6d3ab5d1403a9df9ab420ad' });
      expect(wrapper.state()).toMatchSnapshot();
    });

    test('submit without values', () => {
      const props = {
        isOpen: false,
        isLoading: false,
        verifyDocument: null,
        verifyError: 'error',
        closeVerifyModal: jest.fn(),
        reset: jest.fn(),
        resetError: jest.fn(),
        verifyDocumentRequest: jest.fn(),
        handleSubmit: jest.fn(),
        store: store
      };

      const wrapper = shallow(
        <Verify.WrappedComponent {...props} />
      ).dive().dive().dive();

      wrapper.instance().submit({});
      expect(wrapper.state()).toMatchSnapshot();
    });

  });

  test('simulate close button', () => {
    const props = {
      isOpen: true,
      isLoading: false,
      verifyDocument: {
        "uri": "https://strato-external-storage.s3.amazonaws.com/1530182373708-widescreen.jpeg",
        "timeStamp": null,
        "signers": [
          "6e873015e8ff27d7c6d3ab5d1403a9df9ab420ad",
          "a51f27e78aef85a06631f0725f380001e0ae9fb6"
        ]
      },
      verifyError: 'error',
      closeVerifyModal: jest.fn(),
      reset: jest.fn(),
      resetError: jest.fn(),
      verifyDocumentRequest: jest.fn(),
      handleSubmit: jest.fn(),
      store: store
    };

    const wrapper = shallow(
      <Verify.WrappedComponent {...props} />
    ).dive().dive().dive();

    wrapper.find('Button').simulate('click');
    expect(props.closeVerifyModal).toHaveBeenCalled();
    expect(props.closeVerifyModal).toHaveBeenCalledTimes(1);
  });

  describe('componentWillReceiveProps', () => {
    test('with values', () => {
      const props = {
        isOpen: false,
        isLoading: false,
        verifyDocument: null,
        verifyError: 'error',
        closeVerifyModal: jest.fn(),
        reset: jest.fn(),
        resetError: jest.fn(),
        verifyDocumentRequest: jest.fn(),
        handleSubmit: jest.fn(),
        store: store
      };

      const wrapper = shallow(
        <Verify.WrappedComponent {...props} />
      ).dive().dive().dive();

      wrapper.instance().componentWillReceiveProps({ verifyError: 'error' });
      expect(props.resetError).toHaveBeenCalled();
      expect(props.resetError).toHaveBeenCalledTimes(1);
    });

    test('without values', () => {
      const props = {
        isOpen: false,
        isLoading: false,
        verifyDocument: null,
        verifyError: 'error',
        closeVerifyModal: jest.fn(),
        reset: jest.fn(),
        resetError: jest.fn(),
        verifyDocumentRequest: jest.fn(),
        handleSubmit: jest.fn(),
        store: store
      };

      const wrapper = shallow(
        <Verify.WrappedComponent {...props} />
      ).dive().dive().dive();

      wrapper.instance().componentWillReceiveProps({ verifyError: null });
      expect(props.resetError).not.toHaveBeenCalled();
      expect(props.resetError).not.toHaveBeenCalledTimes(1);
    });
  });

  test('close dialog', () => {
    const props = {
      isOpen: false,
      isLoading: false,
      verifyDocument: null,
      verifyError: 'error',
      closeVerifyModal: jest.fn(),
      reset: jest.fn(),
      resetError: jest.fn(),
      verifyDocumentRequest: jest.fn(),
      handleSubmit: jest.fn(),
      store: store
    };

    const wrapper = shallow(
      <Verify.WrappedComponent {...props} />
    ).dive().dive().dive();

    wrapper.find('Dialog').get(0).props.onClose();
    expect(props.closeVerifyModal).toHaveBeenCalled();
    expect(props.closeVerifyModal).toHaveBeenCalledTimes(1);
  });

  test('mapStateToProps', () => {
    const state = {
      verify: {
        isOpen: false,
        isLoading: false,
        verifyDocument: {
          "uri": "https://strato-external-storage.s3.amazonaws.com/1530182373708-widescreen.jpeg",
          "timeStamp": 1530182371,
          "signers": [
            "6e873015e8ff27d7c6d3ab5d1403a9df9ab420ad",
            "a51f27e78aef85a06631f0725f380001e0ae9fb6"
          ]
        },
        error: null
      }
    }

    expect(mapStateToProps(state)).toMatchSnapshot();
  })

});