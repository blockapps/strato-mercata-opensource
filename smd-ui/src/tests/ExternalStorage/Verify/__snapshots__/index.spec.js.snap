// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify: index form submit with values 1`] = `
Object {
  "errors": Object {},
}
`;

exports[`Verify: index form submit without values 1`] = `
Object {
  "errors": Object {
    "contractAddress": "Can't be blank",
  },
}
`;

exports[`Verify: index mapStateToProps 1`] = `
Object {
  "isLoading": false,
  "isOpen": false,
  "verifyDocument": Object {
    "signers": Array [
      "6e873015e8ff27d7c6d3ab5d1403a9df9ab420ad",
      "a51f27e78aef85a06631f0725f380001e0ae9fb6",
    ],
    "timeStamp": 1530182371,
    "uri": "https://strato-external-storage.s3.amazonaws.com/1530182373708-widescreen.jpeg",
  },
  "verifyError": null,
}
`;

exports[`Verify: index render component with values 1`] = `
"<div>
  <form>
    <Blueprint.Dialog isOpen={true} onClose={[Function: onClose]} iconName=\\"saved\\" title=\\"Valid Resource\\" className=\\"pt-dark verify-dialog\\" canOutsideClickClose={true}>
      <div>
        <div className=\\"pt-dialog-body\\">
          <div className=\\"row content-margin\\">
            <div className=\\"col-sm-3\\">
              <label>
                 URI 
              </label>
            </div>
            <div className=\\"col-sm-9\\">
              <label>
                 
                https://strato-external-storage.s3.amazonaws.com/1530182373708-widescreen.jpeg
                 
              </label>
            </div>
          </div>
          <div className=\\"row content-margin\\">
            <div className=\\"col-sm-3\\">
              <label>
                 Date Uploaded 
              </label>
            </div>
            <div className=\\"col-sm-9\\">
              <label>
                 
                2018-07-16 01:19:24 AM
                 
              </label>
            </div>
          </div>
          <div className=\\"row content-margin verify-result\\">
            <div className=\\"col-sm-3\\">
              <label>
                 Signatures 
              </label>
            </div>
            <div className=\\"col-sm-9\\">
              <ul>
                 
                <li>
                   
                  6e873015e8ff27d7c6d3ab5d1403a9df9ab420ad
                   
                </li>
                <li>
                   
                  a51f27e78aef85a06631f0725f380001e0ae9fb6
                   
                </li>
                 
              </ul>
            </div>
          </div>
        </div>
        <div className=\\"pt-dialog-footer\\">
          <div className=\\"pt-dialog-footer-actions button-center\\">
            <Blueprint.Button intent={0} onClick={[Function: onClick]} text=\\"Close\\" />
          </div>
        </div>
      </div>
    </Blueprint.Dialog>
  </form>
</div>"
`;

exports[`Verify: index render component without values 1`] = `
"<div>
  <form>
    <Blueprint.Dialog isOpen={false} onClose={[Function: onClose]} iconName=\\"pt-icon-info-sign\\" title=\\"Verify\\" className=\\"pt-dark verify-dialog\\" canOutsideClickClose={true}>
      <div className=\\"pt-dialog-body upload-form\\">
        <div className=\\"row\\">
          <div className=\\"col-sm-4 text-right\\">
            <label className=\\"pt-label smd-pad-4\\">
              Contract Address
            </label>
          </div>
          <div className=\\"col-sm-8 smd-pad-4\\">
            <Field name=\\"contractAddress\\" component=\\"input\\" type=\\"text\\" placeholder=\\"Contract Address\\" className=\\"pt-input form-width\\" tabIndex=\\"1\\" required={true} />
            <br />
            <span className=\\"error-text\\" />
          </div>
        </div>
        <div className=\\"pt-dialog-footer\\">
          <div className=\\"pt-dialog-footer-actions button-center smd-margin-8\\">
            <Blueprint.Button intent={0} onClick={[undefined]} disabled={false} text=\\"Attest\\" />
          </div>
        </div>
      </div>
    </Blueprint.Dialog>
  </form>
</div>"
`;
